### DISCUSSION TITLE
운영체제 - 운영체제 구조와 컴퓨터 구조 - 호제&진호

### DISCUSSION THREAD

**Discussion by Jihn0118** [Shortudy-10th/tech-for-developer#96] [질의 응답]
- 커널에 대해 설명하고, 커널의 역할에 대해 말해주세요.
- 하드웨어에서 인터럽트가 발생하는 경우, 운영체제에서 인터럽트를 처리하는 과정에 대해 설명해주세요.
- 컴퓨터의 기본적인 구성 요소에 대해 설명해주세요.
- 운영체제의 중요한 점은 "다중 프로그래밍 (Multi programming)"이 가능하다는 점입니다. 다중 프로그래밍이란 무엇인가요?
- 명령어 처리 과정에 대해 설명해주세요.

**Comment by Jihn0118** [comment]
## ❓️ 커널에 대해 설명하고, 커널의 역할에 대해 말해주세요.

💬 답변

커널은 운영체제의 핵심 부분으로, 기본적인 시스템 작업을 관리하고 하드웨어 리소스를 제어한다.

커널은 사용자 프로그램과 하드웨어 사이에서 인터페이스 역할을 하면서, 메모리 관리, 프로세스 관리, 파일 시스템 관리, 장치 드라이버 관리 등 다양한 역할을 수행한다. 

커널은 시스템의 안정성과 효율성을 높이는 데 있어 중요한 역할을 담당한다.

**Comment by Jihn0118** [comment]
## ❓️ 하드웨어에서 인터럽트가 발생하는 경우, 운영체제에서 인터럽트를 처리하는 과정에 대해 설명해주세요.

💬 답변

_인터럽트는 CPU가 현재 실행 중인 작업을 잠시 멈추고, 우선순위가 높은(IO 장치의 완료 신호 등) 작업을 처리한 후 다시 원래의 작업을 진행하도록 해준다._

### 인터럽트가 발생하는 경우 운영체제에서 처리하는 과정

1. 현재 실행 중인 작업의 상태를 저장하고, 현재 프로세스의 컨텍스트를 보호하기 위해 CPU 내 레지스터 값들을 스택에 `push`하여 저장한다.
2. 인터럽트 벡터에서 해당 인터럽트 처리를 위한 인터럽트 핸들러의 주소를 구한다.
3. 구한 인터럽트 핸들러의 주소로 이동하여 핸들러 코드를 실행한다. 이후 인터럽트 핸들러는 인터럽트의 원인을 파악하고 해당 이벤트를 처리한 후 인터럽트를 종료한다.
4. 인터럽트 처리가 끝나면, 스택에 보관한 CPU 레지스터 값을 `pop`하여 복원하고 원래 작업을 이어서 진행하기 위해 저장했던 상태로 되돌린다.

인터럽트는 운영체제에서 가장 중요한 작업 중 하나로, 특별한 작업을 필요로 하거나 시스템 상태 변화에 빠르게 대응할 수 있는 매커니즘을 제공한다.

**Comment by Jihn0118** [comment]
## ❓️ 컴퓨터의 기본적인 구성 요소에 대해 설명해주세요.

💬 답변

![image](https://github.com/Shortudy-10th/tech-for-developer/assets/83208807/8cfd6d0d-707d-49f4-aead-486ed4e144e8)

컴퓨터 시스템은 하드웨어와 소프트웨어로 구성된다. 하드웨어에는 중앙처리장치(CPU), 기억 장치, 입•출력장치(I•O)가 있다. 

### CPU (중앙 처리 장치)와 메모리
- 프로그램을 실행하고 데이터를 처리하는 기능을 수행한다.
- 연산 장치와 제어 장치로 구성된다.

#### 주기억장치
- CPU가 직접 접근 가능하다.
- 현재 수행되는 프로그램과 데이터를 저장한다.
- 영구 저장 능력이 없어서 일시적 저장 장치로 사용한다.
  - RAM : 자유롭게 읽고 쓸 수 있으며, 전원이 꺼지면 저장된 내용이 사라지는 휘발성 메모리
  - ROM : 읽을 수만 있는 기억장치로, 전원이 꺼져도 저장된 내용이 사라지지 않는다.
#### 보조기억장치
- 비 휘발성 메모리
- 용량이 크지만 읽기/쓰기 속도가 주기억장치보다 느리다.
- 하드디스크, CD, DVD 등

### 입출력장치
- 입력 장치는 외부의 데이터를 컴퓨터로 입력하는 장치이다.
- 출력 장치는 컴퓨터에서 처리한 결과를 사용자가 원하는 형태로 출력하는 장치이다.

### 저장장치
- 메모리보다 느리지만 저렴하고 용량이 크며, 전원 온/오프 상관없이 데이터를 영구적으로 저장한다.

### 메인보드
- CPU와 메모리 등 다양한 부품을 연결하는 커다란 판이다.
- 버스가 복잡하게 얽혀있어 부품들이 작동할 수 있다.

**Comment by Jihn0118** [comment]
## ❓️ 다중 프로그래밍(Multi-programming)이란 무엇인가요?

💬 답변

<다중 프로그래밍의 탄생 배경>
초기 컴퓨터는 하나의 프로세서가 **하나의 프로세스만 처리할 수 있도록 설계**되었다. 
하지만 하나의 프로세스를 처리하는 과정에서 프로세서의 처리 속도와 입출력 속도 간의 차이로 인해 **입출력 처리가 완료될 때까지 기다리는 비효율적인 상황** 이 발생하게 된다.

## 다중 프로그래밍
![zxc](https://github.com/Shortudy-10th/tech-for-developer/assets/87293724/59f740fd-7abf-4e41-82f2-cdc081989d7b)

- 다중 프로그래밍(Multiprogramming)은 컴퓨터 시스템의 운영 방식 중 하나로, **여러 개의 프로세스가 동시에 메모리에 적재되어 실행되는 방식**을 나타낸다.
- CPU와 다른 시스템 자원을 보다 **효율적으로 활용하여 처리 능력을 높이는 것을 목표**로 한다.
- 주요 목표는 CPU를 최대한 유지하고, **입출력 작업 등으로 인한 Blocking 상황에서도 시스템 자원을 효율적으로 사용**하는 것이다.
- 이로 인해 시스템 **전체적인 처리량이 증가**하며, 사용자 또는 응용 프로그램의 **대기 시간이 줄어든다.**

**Reply by zini9188** [reply]
❓️ 다중 프로그래밍이 사용되는 현대 컴퓨터 시스템의 대표적인 예시는 어떤 것들이 있나요?

**Reply by Jihn0118** [reply]
1. 개인 컴퓨터 및 노트북
 - 사용자가 여러 개의 응용 프로그램을 동시에 실행하거나 여러 탭 또는 창에서 다양한 작업을 수행할 수 있도록 운영체제가 이러한 작업들을 조율하여 빠르게 전환하며 다중 프로그래밍의 이점을 활용합니다.

2. 서버 및 데이터 센터
 - 여러 클라이언트의 요청을 동시에 처리할 수 있게 웹 서버, DB 서버, 클라우드 컴퓨팅 환경 등에서 다수의 프로세스가 동시에 실행되어 다양한 작업을 처리하며 시스템의 자원 활용도를 최대화합니다.

3. 스마트폰 및 태블릿
 - 스마트폰과 태블릿은 다중 프로그래밍을 활용하여 사용자가 다양한 애플리케이션을 백그라운드에서 실행할 수 있도록 합니다. 
 메시지 앱, 이메일 앱, 브라우저, 음악 앱 등 여러 개의 애플리케이션을 동시에 관리하여 사용자 경험을 향상시킵니다.


이 외에도 거의 모든 컴퓨팅 환경에서 다중 프로그래밍이 적용되어 시스템의 효율성을 향상 시키고 여러 작업을 동시에 처리할 수 있도록 도와줍니다.

**Comment by Jihn0118** [comment]
## ❓️ 명령어 처리 과정에 대해 설명해주세요.

💬 답변
1. 프로그램 로드(Load):
사용자 또는 응용 프로그램이 실행할 프로그램이나 명령어들이 메모리에 로드된다. 이때 운영체제는 프로그램의 실행 시작 주소를 알고 있어야 합니다.

2. 명령어 가져오기 (Fetch):
CPU는 프로그램 카운터(Program Counter, PC)에 저장된 주소로부터 다음 실행할 명령어를 메모리에서 가져옵니다. 프로그램 카운터는 현재 실행 중인 명령어의 주소를 가리킵니다.

3. 명령어 해석 (Decode):
가져온 명령어는 명령어 레지스터(Instruction Register, IR)에 저장되며, CPU는 이 명령어를 해석합니다. 명령어는 어떤 작업을 수행해야 하는지, 어떤 레지스터나 메모리 위치를 사용해야 하는지 등의 정보를 담고 있습니다.

4. 피연산자 획득 (Fetch Operands):
명령어의 실행에 필요한 피연산자(데이터)를 메모리에서 가져오거나 레지스터에서 읽어옵니다. 이 단계에서 메모리 접근이 필요한 경우 캐시 메모리 등을 통해 효율적인 데이터 접근을 시도합니다.

5. 연산 수행 (Execute):
명령어에 지정된 연산을 수행합니다. 이 단계에서 CPU는 레지스터를 사용하여 산술 연산, 논리 연산 등을 수행합니다.

6. 결과 저장 (Write Back):
연산의 결과를 레지스터나 메모리에 저장합니다. 이 단계에서 연산 결과가 다음 명령어나 응용 프로그램에서 사용될 수 있도록 준비됩니다.

7. 프로그램 카운터 갱신 (Update Program Counter):
다음에 실행할 명령어를 가리키는 프로그램 카운터(PC)를 업데이트하여 다음 명령어의 주소로 이동합니다. 이를 통해 다음 명령어를 가져와서 처리하게 됩니다.

이러한 명령어 처리 과정은 컴퓨터 시스템의 핵심 동작이며, CPU의 구조와 아키텍처, 운영체제의 역할에 따라 상세 내용이 달라질 수 있습니다.
